File: web/src/types/catalog.ts
import type { Agent, ModelConfig } from "./agent";

export type CatalogCategory =
  | "coding"
  | "research"
  | "writing"
  | "productivity"
  | "marketing"
  | "data"
  | "ops"
  | "other";

export interface CatalogAgentSummary {
  id: string;
  slug: string;
  name: string;
  description?: string;
  icon?: string;
  category: string;
  owner_email: string;
  current_version_number: number;
  current_version_id: string;
  mcp_servers: string[];
  import_count: number;
  is_published: boolean;
  unpublished_at?: string | null;
  updated_at: string;
}

export interface McpToolBindingSnapshot {
  server_name: string;
  tool_name: string;
}

export interface AgentVersion {
  id: string;
  catalog_agent_id: string;
  version_number: number;
  name: string;
  description?: string;
  icon?: string;
  instruction: string;
  model_config: ModelConfig;
  tools: string[];
  mcp_servers: unknown[];
  mcp_tool_bindings: McpToolBindingSnapshot[];
  category: string;
  release_notes?: string | null;
  published_by: string;
  published_at: string;
  unpublished_at?: string | null;
}

export interface AgentVersionSummary {
  id: string;
  version_number: number;
  release_notes?: string | null;
  published_at: string;
  unpublished_at?: string | null;
}

export interface CatalogAgentDetail {
  id: string;
  slug: string;
  owner_id: string;
  owner_email: string;
  is_published: boolean;
  unpublished_at?: string | null;
  current_version_id?: string | null;
  current_version_number?: number | null;
  created_at: string;
  updated_at: string;
  import_count: number;
  version: AgentVersion;
  imported_as_agent_id?: string | null;
}

export interface CategoryFacet {
  name: string;
  count: number;
}

export interface ServerFacet {
  name: string;
  count: number;
}

export interface CatalogFacets {
  categories: CategoryFacet[];
  servers: ServerFacet[];
}

export interface CatalogListResponse {
  items: CatalogAgentSummary[];
  total: number;
}

export interface AgentVersionListResponse {
  items: AgentVersionSummary[];
  total: number;
}

export interface MyCatalogAgent {
  id: string;
  slug: string;
  name: string;
  description?: string;
  icon?: string;
  category: string;
  is_published: boolean;
  unpublished_at?: string | null;
  current_version_number?: number | null;
  linked_agent_id?: string | null;
  import_count: number;
  created_at: string;
  updated_at: string;
}

export interface MyCatalogListResponse {
  items: MyCatalogAgent[];
  total: number;
}

export interface ImportRequest {
  items: Array<{
    catalog_agent_id: string;
    pinned_version_id?: string | null;
  }>;
}

export interface ImportResponse {
  created: Agent[];
  already_imported: Agent[];
}

export interface DriftResponse {
  is_dirty: boolean;
  latest_version_number?: number | null;
  latest_version_id?: string | null;
  changed_fields: string[];
}

File: api/app/schemas/catalog.py
"""Catalog-related Pydantic schemas."""

from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field

from app.schemas._common import MODEL_CONFIG_ALIAS, OptionalUuidStr, UuidStr


class CatalogCategory(str, Enum):
    """Controlled vocabulary for catalog filtering — can grow without DB migration
    because storage side uses VARCHAR(32)."""

    coding = "coding"
    research = "research"
    writing = "writing"
    productivity = "productivity"
    marketing = "marketing"
    data = "data"
    ops = "ops"
    other = "other"


class PublishRequest(BaseModel):
    category: CatalogCategory
    release_notes: str | None = None


class DriftResponse(BaseModel):
    is_dirty: bool
    latest_version_number: int | None = None
    latest_version_id: OptionalUuidStr = None
    changed_fields: list[str] = []


class McpToolBinding(BaseModel):
    server_name: str
    tool_name: str


class AgentVersionResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True, populate_by_name=True, protected_namespaces=(),
    )

    id: UuidStr
    catalog_agent_id: UuidStr
    version_number: int
    name: str
    description: str | None = None
    icon: str | None = None
    instruction: str
    model_config_json: dict = Field(**MODEL_CONFIG_ALIAS)
    tools: list
    mcp_servers: list
    mcp_tool_bindings: list[McpToolBinding]
    category: str
    release_notes: str | None = None
    published_by: UuidStr
    published_at: datetime
    unpublished_at: datetime | None = None


class AgentVersionSummary(BaseModel):
    """Lightweight version descriptor for history lists."""

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: UuidStr
    version_number: int
    release_notes: str | None = None
    published_at: datetime
    unpublished_at: datetime | None = None


class AgentVersionListResponse(BaseModel):
    items: list[AgentVersionSummary]
    total: int


class CatalogAgentSummary(BaseModel):
    """Shape used in the /catalog list view — hydrated from JOIN with the
    current AgentVersion snapshot."""

    model_config = ConfigDict(populate_by_name=True, protected_namespaces=())

    id: UuidStr
    slug: str
    name: str
    description: str | None = None
    icon: str | None = None
    category: str
    owner_email: str
    current_version_number: int
    current_version_id: UuidStr
    mcp_servers: list[str] = []  # unique server names extracted from mcp_tool_bindings
    import_count: int = 0
    is_published: bool
    unpublished_at: datetime | None = None
    updated_at: datetime


class CatalogListResponse(BaseModel):
    items: list[CatalogAgentSummary]
    total: int


class CatalogAgentDetail(BaseModel):
    """Full detail for the catalog detail page — combines CatalogAgent
    identity with the current (or requested) AgentVersion snapshot."""

    model_config = ConfigDict(populate_by_name=True, protected_namespaces=())

    id: UuidStr
    slug: str
    owner_id: UuidStr
    owner_email: str
    is_published: bool
    unpublished_at: datetime | None = None
    current_version_id: OptionalUuidStr = None
    current_version_number: int | None = None
    created_at: datetime
    updated_at: datetime
    import_count: int = 0
    # The effective version being viewed (may be latest or a specific version
    # when the request asked for one).
    version: AgentVersionResponse
    # Whether the requester currently has an import for this catalog agent.
    imported_as_agent_id: OptionalUuidStr = None


class CategoryFacet(BaseModel):
    name: str
    count: int


class ServerFacet(BaseModel):
    name: str
    count: int


class CatalogFacetsResponse(BaseModel):
    categories: list[CategoryFacet]
    servers: list[ServerFacet]


# ── Imports ──────────────────────────────────────────────────────────────

class ImportItem(BaseModel):
    catalog_agent_id: UuidStr
    pinned_version_id: OptionalUuidStr = None


class ImportRequest(BaseModel):
    items: list[ImportItem]


class ImportPatchRequest(BaseModel):
    pinned_version_id: OptionalUuidStr = None


# ── Publisher-facing ─────────────────────────────────────────────────────

class RollbackRequest(BaseModel):
    version_id: UuidStr


class MyCatalogAgent(BaseModel):
    """An entry in GET /catalog/mine."""

    model_config = ConfigDict(populate_by_name=True, protected_namespaces=())

    id: UuidStr
    slug: str
    name: str
    description: str | None = None
    icon: str | None = None
    category: str
    is_published: bool
    unpublished_at: datetime | None = None
    current_version_number: int | None = None
    linked_agent_id: OptionalUuidStr = None
    import_count: int = 0
    created_at: datetime
    updated_at: datetime


class MyCatalogListResponse(BaseModel):
    items: list[MyCatalogAgent]
    total: int
